dotnet/src/SemanticKernel.Abstractions/Filters/Function/FunctionInvocationContext.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.SemanticKernel;
+
+/// <summary>
+/// Class with data related to function invocation.
+/// </summary>
+[Experimental("SKEXP0001")]
+public class FunctionInvocationContext
+{
+    /// <summary>
+    /// Initializes a new instance of the <see cref="FunctionInvocationContext"/> class.
+    /// </summary>
+    /// <param name="function">The <see cref="KernelFunction"/> with which this filter is associated.</param>
+    /// <param name="arguments">The arguments associated with the operation.</param>
+    internal FunctionInvocationContext(KernelFunction function, KernelArguments arguments)
+    {
+        Verify.NotNull(function);
+        Verify.NotNull(arguments);
+
+        this.Function = function;
+        this.Arguments = arguments;
+    }
+
+    /// <summary>
+    /// Gets the <see cref="KernelFunction"/> with which this filter is associated.
+    /// </summary>
+    public KernelFunction Function { get; }
+
+    /// <summary>
+    /// Gets the arguments associated with the operation.
+    /// </summary>
+    public KernelArguments Arguments { get; }
+
+    /// <summary>
+    /// Gets a dictionary of metadata associated with the operation.
+    /// </summary>
+    public IReadOnlyDictionary<string, object?>? Metadata { get; internal set; }
+
+    /// <summary>
+    /// Gets the result of the function's invocation.
+    /// </summary>
+    public FunctionResult? Result { get; internal set; }
+
+    /// <summary>
+    /// Gets the raw result of the function's invocation.
+    /// </summary>
+    internal object? ResultValue { get; private set; }
+
+    /// <summary>
+    /// Sets an object to use as the overridden new result for the function's invocation.
+    /// </summary>
+    /// <param name="value">The value to use as the new result of the function's invocation.</param>
+    public void SetResultValue(object? value)
+    {
+        this.ResultValue = value;
```

Comment:
It's strange that calling SetResultValue won't impact Result.

Reply:
Yeah, this is how result overriding is designed in today's version of SK, I think we should redesign it as well. Maybe we should allow to set entire `Result` object.

Reply:
`SetResultValue` was removed and `Result` property is now available with public setter. Usage example:
```csharp
context.Result = new FunctionResult(context.Function, "Friendly message instead of exception");
```

Reply:
Using the code above has the side effect of removing the current `culture` and `metadata`. Developers will need to take care to avoid this, the old pattern didn't have this downside.
One other thing to consider is what happens if we add a new property to `FunctionResult`. Existing filter implementations would need to be updated to prevent data loss.

Would `context.Result = new FunctionResult(context.Result, "Friendly message instead of exception");` be a better pattern?

Reply:
> Would `context.Result = new FunctionResult(context.Result, "Friendly message instead of exception");` be a better pattern?

I applied this pattern. `FunctionResult` has new constructor that accepts instance of `FunctionResult` to copy data from and additional `value` parameter to override result value.

> We should also consider obsoleting the current `FunctionResult` constructor in favour of init properties.

We can keep existing constructor without obsoleting it, but make every new property with `init` capability. I updated `Function`, `Metadata` and `Culture` properties with `init`. I also changed the visibility of `Culture` property from `internal` to `public`, since it's possible to set it via constructor of `FunctionResult`, but after that users don't have access to it, which is a little bit confusing. 

With latest updates it's possible to modify `Culture` and `Metadata` inside filter by creating new instance of `FunctionResult` and I've added more tests and examples to verify that it works correctly.